### PR TITLE
[PW-6021] Add 'maintain state' option on payment confirmation

### LINKED ISSUE #1276: [PW-6021] Forced order status change for Capture notifications
**Describe the bug**
Force change of order status on capture notifications

Due to the capture of payment being one of the last functions in our order flow (we don't charge customer card until after despatch)
we require the order status to remain in its last state/status and not be forced back into a "processing" state / status.

This was possible in 7.2.0 by not allocating a dropdown option in admin setting 
"Order status: payment confirmed" --Please Select--

Since update if a option is not selected it results in a order status update to the default "closed" state in magento

Can you allow for a option of "No Status Change" 

I don't really see the requirement to restrict the order status options to only a Processing state either.  
A manual capture request could be requested during a On Hold or Complete state.

**To Reproduce**
Updated from 7.2.0 - 7.3.2
Suspect this may be a issue related to changes made to resolve [PW-5812]

**Expected behavior**
No change of order status if one is not selected in admin dropdown options

**Magento version**
2.4.3

**Plugin version**
7.3.2

### PR DESCRIPTION
**Pull request by tnaber**
**Description**
In #1276 it was suggested to implement a 'No status change' option on payment confirmation to allow for different flows (e.g. capturing after POD, see discussion). This was already possible in 7.2.0 by not setting any status (leaving it to '-- Please Select --'), however this feature disappeared for later versions. This PR adds the option to select 'Maintain status' in the `Order status: payment confirmed` dropdown. The '-- Please Select --' option has been removed to prevent confusion to which status will be selected, the default is 'Processing'.

The state will be maintained only in the Manual Capture flow.

image
 
**Tested scenarios**
- [x] Manual capture
- [x] Immediate capture
- [x] Verified by community

**Fixed issue**: 
Closes #1276

### CODE REVIEW

File: Model/Config/Source/Status/ProcessingMaintain.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+/**
+ * Order Statuses source model
+ */
+class ProcessingMaintain extends \Magento\Sales\Model\Config\Source\Order\Status
+{
+    /**
+     * @var string[]
+     */
+    protected $_stateStatuses = [
```
Comment (Morerice) [inline Model/Config/Source/Status/ProcessingMaintain.php:18]: Could you set these data members to private and remove the underscore?

File: Helper/Webhook.php [hunk 2]
```diff
@@ -70,6 +70,9 @@
 
 class Webhook
 {
+    // Adyen Custom States
+    const STATE_MAINTAIN = ['maintain' => 'Maintain status'];
```
[the remark this answers is not in the document]
Comment (acampos1916) [reply Helper/Webhook.php]: @tnaber resolving this comment as you already addressed it

File: Helper/Webhook.php [hunk 3]
```diff
@@ -1623,7 +1633,12 @@ private function finalizeOrder(Order $order, Notification $notification)
             $comment = "Adyen Payment Successfully completed";
             // If a status is set, add comment, set status and update the state based on the status
             // Else add comment
-            if (!empty($status)) {
+            if (!empty($status) && $maintainingState) {
+                $order->addStatusHistoryComment(__($comment), $status);
+                $this->logger->addAdyenNotificationCronjob(
```
[the remark this answers is not in the document]
Comment (acampos1916) [reply Helper/Webhook.php:1636]: @tnaber resolving this comment as you already addressed it

File: Helper/Webhook.php [hunk 4]
```diff
@@ -1572,6 +1575,13 @@ private function finalizeOrder(Order $order, Notification $notification)
             $order->getStoreId()
         );
 
+        // Set state back to previous state to prevent update if 'maintain status' was configured
+        $maintainingState = false;
+        if ($status == key(self::STATE_MAINTAIN)) {
```
[the remark this answers is not in the document]
Comment (acampos1916) [reply Helper/Webhook.php]: @tnaber this one still needs resolving

File: Model/Config/Source/Status/ProcessingMaintain.php (hunk 1 above)
Comment (acampos1916) [reply Model/Config/Source/Status/ProcessingMaintain.php]: @tnaber this one still needs resolving

File: Model/Config/Source/Status/AdyenStates.php [hunk 5]
```diff
@@ -0,0 +1,11 @@
+<?php
+
+
+namespace Adyen\Payment\Model\Config\Source\Status;
+
+
+class AdyenStates
```
Comment (Morerice) [inline Model/Config/Source/Status/AdyenStates.php:7]: Can you rename it to the singular `AdyenState` and add a new line at the end of the file?

File: Model/Config/Source/Status/ProcessingMaintain.php [hunk 6]
```diff
@@ -0,0 +1,50 @@
+<?php
```
Comment (Morerice) [inline Model/Config/Source/Status/ProcessingMaintain.php:1]: Can you add the Adyen signature?

**Review by Morerice — CHANGES_REQUESTED** [review]

**Review by Morerice — APPROVED** [review]

**Review by acampos1916 — APPROVED** [review]

**Comment by tnaber** [comment]
Hi @sheppeybrian,

Sad to hear that you had to go back to 7.2.0 again. I don't completely understand what you mean so I'll write down the steps I'm taking, can you point out where your flow differs from mine?

My setup:
image

1. Place order, order in Payment Review
2. Authorization notification received, order in Pending
3. Invoice created, order in Pending
4. Capture notification received, order in Pending
5. Ship order, order Complete

When testing the flow where we ship before creating the invoice I notice some strange behaviour
1. Place order, order in Payment Review
2. Authorization notification received, order in Pending
3. Order shipped, order in Processing
4. Invoice created, order Closed
5. Capture notification received but not processed, order in Closed

Am I correct in assuming that you are using the second flow? 
If so, do you mean that an invoice should be directly opened as a 'Paid' invoice? As far as I understand it, the capture notification has not been received at that point

If this is not the flow you mean, would you able to provide your flow in a similar format?

Cheers,
Titus

**Comment by sheppeybrian** [comment]
Hello Titus

So our flow is the same as your second example except that 3 & 4 are
reversed.  I don't see this making much of a difference.

However it's the Invoice created stage that's the issue. (4)
It looked fine during testing as the invoice is created, but on closer
inspection the invoice status is marked "Status:Pending" and not "Paid"
Due to this, Magento will try to prevent you from handling the order any
further and close it

So I would deem that magento is correctly issuing an invoice as "pending"
awaiting a response for Adyen "capture" to update it to "Paid", which does
not occur.

I'm not a developer but this is what I see as being the issue that
occurred.


Many Thanks


Brian



AstroTex Ltd t/a expressgrass.com
Tribune Drive, Trinity Trading Estate
SITTINGBOURNE
Kent
ME10 2PG Map <https://www.google.com/maps/dir//expressgrass>

*tel*: 0844 855 4704
*direct line*: 01795 385202
*fax*: 01795 385001
*email*:
***@***.*** ***@***.***>**web*: www.expressgrass.com
*Reviews: Google Reviews <https://g.page/expressgrass/review?kd>*

P please don't print this e-mail unless you really need to

This email and its attachments may be confidential and are intended solely
for the use of the individual to whom it is addressed. Any views or
opinions expressed are solely those of the author and do not necessarily
represent those of expressgrass.com. If you are not the intended recipient
of this email and its attachments, you must take no action based upon them,
nor must you copy or show them to anyone. Please contact the sender if you
believe you have received this email in error

**Comment by tnaber** [comment]
Hi Brian,
I did not expect this behaviour to occur and will be diving back into this :)

**Comment by sheppeybrian** [comment]
No worries, thanks for your help, we did not notice the issue until we put
it on the live site and started to process orders.


Brian

**Comment by jordanvector** [comment]
@tnaber was there ever a follow up ticket created to track the fix for the invoice status issue referenced above?

**Comment by tnaber** [comment]
Hi @jordanvector,
A follow up ticket has been created and the investigation is in progress. 
The issue is not necessarily caused by this PR, the problem exists since orders are pushed to the Processing state after an invoice in order to allow for multiple invoices. If an order is shipped before being captured, it will move to Closed. We are currently figuring out the correct flow.

Cheers,
Titus

**Comment by jordanvector** [comment]
@tnaber understood thank you for the update.

**Comment by tnaber** [comment]
Hi @jordanvector and @sheppeybrian,

In #1533 you can view our proposed solution. Let me know what you think!
Also let me know whether you think the 'Maintain state' option has become redundant due to this (as the shipping before capture flow can be used without it)

Cheers,
Titus

**Pull request merged**